subprojects/core/src/main/java/org/gradle/internal/classpath/DefaultCachedClasspathTransformer.java

```diff
@@ -161,22 +163,55 @@ public List<URL> transform(Collection<URL> urls, StandardTransform transform) {
         if (urls.isEmpty()) {
             return ImmutableList.of();
         }
+        TypeHierarchyRegistry typeRegistry = getTypeRegistry(urls);
         ClasspathFileTransformer transformer = fileTransformerFor(transform);
         return transformAll(
             urls,
-            (url, seen) -> cachedURL(url, transformer, seen)
+            (url, seen) -> cachedURL(url, transformer, seen, typeRegistry)
         );
     }
 
     private ClassPath transformFiles(ClassPath classPath, ClasspathFileTransformer transformer) {
+        TypeHierarchyRegistry typeRegistry = getTypeRegistry(classPath.getAsFiles());
         return DefaultClassPath.of(
             transformAll(
                 classPath.getAsFiles(),
-                (file, seen) -> cachedFile(file, transformer, seen)
+                (file, seen) -> cachedFile(file, transformer, seen, typeRegistry)
             )
         );
     }
 
+    private TypeHierarchyRegistry getTypeRegistry(Collection<URL> urls) {
+        List<File> files = urls.stream()
+            .filter(url -> url.getProtocol().equals("file"))
+            .map(Convert::urlToFile)
+            .collect(Collectors.toList());
+        return getTypeRegistry(files);
+    }
+
+    private TypeHierarchyRegistry getTypeRegistry(List<File> files) {
```

Comment:
I would suggest extracting this logic to another class, so that it's not mixed with the transforming & caching stuff.

Reply:
I extracted to the inner class of `DefaultCachedClasspathTransformer` for now, what do you think?

Reply:
LGTM now. 👍